### PR TITLE
[generator_interior] Be more precise with scopes of borrowed places

### LINKED ISSUE #57017: async/await: awaiting inside a match block captures borrow too eagerly
If you use `await!(some_fut)` inside an arm of a `match X`, the generated future eagerly borrows the value of `X`, if it not needed.

This may not usually be noticeable, but the issue compounds when the type `X` contains a trait object, and the future you wish to return is `impl Future + Send`. This causes a misleading error message that "`dyn Trait + Send` cannot be shared between threads", which is required to for `&X: Send`.

## Example

Here's a simple struct with a trait object:

```rust
struct Client(Box<Any + Send>);
```

Consider a function like this:

```rust
impl Client {
    fn status(&self) -> u16 {
        200
    }
}
```

You could consider using a match to determine what kind of future to await (or what arguments to pass):

```rust
async fn get() {
}

pub fn wat() -> impl Future + Send {
    let client = Client(Box::new(true));
    async move {
        match client.status() {
            200 => {
                let _x = await!(get());
            },
            _ => (),
        }
    }
}
```

If the `await` is moved out of the match block, all is well:

```rust
pub fn ok() -> impl Future + Send {
    let client = Client(Box::new(true));
    async move {
        if client.status() == 200 {
            let _x = await!(get());
        }
    }
}
```

The `wat` function causes this compilation error:

```
error[E0277]: `(dyn std::any::Any + std::marker::Send + 'static)` cannot be shared between threads safely
  --> src/main.rs:21:17
   |
21 | pub fn wat() -> impl Future + Send {
   |                 ^^^^^^^^^^^^^^^^^^ `(dyn std::any::Any + std::marker::Send + 'static)` cannot be shared between threads safely
   |
   = help: the trait `std::marker::Sync` is not implemented for `(dyn std::any::Any + std::marker::Send + 'static)`
   = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<(dyn std::any::Any + std::marker::Send + 'static)>`
   = note: required because it appears within the type `std::boxed::Box<(dyn std::any::Any + std::marker::Send + 'static)>`
   = note: required because it appears within the type `Client`
   = note: required because of the requirements on the impl of `for<'r> std::marker::Send` for `&Client`
   = note: required because it appears within the type `for<'r> {Client, &'r Client, u16, impl std::future::Future, ()}`
   = note: required because it appears within the type `[static generator@src/main.rs:23:16: 30:6 client:Client for<'r> {Client, &'r Client, u16, impl std::future::Future, ()}]`
   = note: required because it appears within the type `std::future::GenFuture<[static generator@src/main.rs:23:16: 30:6 client:Client for<'r> {Client, &'r Client, u16, impl std::future::Future, ()}]>`
   = note: required because it appears within the type `impl std::future::Future`
   = note: the return type of a function must have a statically known size
```

Playground: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=2a9dbea32d31457d50d40b99c52ee214 (updated to latest syntax -Niko)

### PR DESCRIPTION
**Pull request by eholk**
Previously the generator interior type checking analysis would use the nearest temporary scope as the scope of a borrowed value. This ends up being overly broad for cases such as:

```rust
fn status(_client_status: &Client) -> i16 {
    200
}

fn main() {
    let client = Client;
    let g = move || match status(&client) {
        _status => yield,
    };
    assert_send(g);
}
```

In this case, the borrow `&client` could be considered in scope for the entirety of the `match` expression, meaning it would be viewed as live across the `yield`, therefore making the generator not `Send`.

In most cases, we want to use the enclosing expression as the scope for a borrowed value which will be less than or equal to the nearest temporary scope. This PR changes the analysis to use the enclosing expression as the scope for most borrows, with the exception of borrowed RValues which are true temporary values that should have the temporary scope. There's one further exception where borrows of a copy such as happens in autoref cases also should be ignored despite being RValues.

Joint work with @nikomatsakis 

Fixes #57017

r? @tmandry

### CODE REVIEW

File: compiler/rustc_typeck/src/check/generator_interior.rs [hunk 1]
```diff
@@ -369,7 +369,17 @@ impl<'a, 'tcx> Visitor<'tcx> for InteriorVisitor<'a, 'tcx> {
 
         self.expr_count += 1;
 
-        let scope = self.region_scope_tree.temporary_scope(expr.hir_id.local_id);
+        debug!("is_borrowed_temporary: {:?}", self.drop_ranges.is_borrowed_temporary(expr));
+
+        let scope = if self.drop_ranges.is_borrowed_temporary(expr) {
```
Comment (nikomatsakis) [inline compiler/rustc_typeck/src/check/generator_interior.rs:374]: This probably needs a comment...

---

Typically, the value produced by an expression is consumed by its parent in some way, so we only have to check if the parent contains a yield (note that the parent may, for example, store the value into a local variable, but then we already consider local variables to be live across their scope). 

However, in the case of temporary values, we are going to store the value into a temporary on the stack that is live for the current temporary scope and then return return a reference to it. That value may be live across the entire temporary scope.

File: compiler/rustc_typeck/src/expr_use_visitor.rs [hunk 2]
```diff
@@ -45,8 +45,17 @@ pub trait Delegate<'tcx> {
...
         diag_expr_id: hir::HirId,
         bk: ty::BorrowKind,
+        is_autoref: bool,
     );
 
+    /// The value found at `place` is being copied.
+    /// `diag_expr_id` is the id used for diagnostics (see `consume` for more details).
+    fn copy(&mut self, place_with_id: &PlaceWithHirId<'tcx>, diag_expr_id: hir::HirId) {
+        // In most cases, treating a copy as a borrow is the right thing, so we forward
+        // this to the borrow callback by default.
```
Comment (nikomatsakis) [inline compiler/rustc_typeck/src/expr_use_visitor.rs:55]: Let's explain why --

"In most cases, copying data from `x` is equivalent to doing `*&x`, so treat this as a borrow of `x`."

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: compiler/rustc_typeck/src/check/generator_interior/drop_ranges/record_consumed_borrow.rs [hunk 3]
```diff
@@ -94,11 +99,40 @@ impl<'tcx> expr_use_visitor::Delegate<'tcx> for ExprUseDelegate<'tcx> {
...
+
         self.places
             .borrowed
             .insert(TrackedValue::from_place_with_projections_allowed(place_with_id));
+
+        // Keep track of whether this is a borrowed temporary (i.e. a borrow of an RValue)
+        // so that later in generator_interior we can use the correct scope.
+        //
+        // We ignore borrows that are the result of an autoref because these will be
+        // immediately consumed and should not extend the temporary's lifetime.
```
[the remark this answers is not in the document]
Comment (eholk) [reply compiler/rustc_typeck/src/check/generator_interior/drop_ranges/record_consumed_borrow.rs:117]: The issue-57017.rs test included in this PR fails without the `is_autoref` check. Here's a trace: https://gist.github.com/eholk/82098d01e812cf30c4051f1f4d87be29

The relevant events are the copy on line 45 which was handled by introducing the copy callback, and the borrow on line 50 which is handled by the `is_autoref` check.

I considered using something different, like checking whether the borrow included a `Deref` projection but I don't have a good argument for why that is more or less correct than checking `is_autoref`.

File: compiler/rustc_typeck/src/check/generator_interior.rs (hunk 1 above)
Comment (eholk) [reply compiler/rustc_typeck/src/check/generator_interior.rs:374]: Done

File: compiler/rustc_typeck/src/expr_use_visitor.rs (hunk 2 above)
Comment (eholk) [reply compiler/rustc_typeck/src/expr_use_visitor.rs:55]: Done

File: compiler/rustc_typeck/src/check/generator_interior/drop_ranges/record_consumed_borrow.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (eholk) [reply compiler/rustc_typeck/src/check/generator_interior/drop_ranges/record_consumed_borrow.rs]: We talked about this today and decided that the right thing to do was to check for `Deref` projections. I've updated the code to do that, and we included a length comment explaining why.

**Comment by tmandry** [comment]
The change makes sense to me and is well documented. Thanks!

@bors r+

**Pull request merged**